### PR TITLE
 Create OptimizerExecutionFrame for graph optimization

### PR DESCRIPTION
**Pull request by weixingzhang**
Similar to PR#442, the idea of this PR is to make CPU kernels easily leveraged by graph optimization.

For some graph optimizations (e.g. conv+bn fusion, constant folding), the computation needs to be done on model parameters to generate new parameters. To achieve it, the better way is to utilize CPU kernels rather than creating new compute functions.

### CODE REVIEW

File: onnxruntime/core/optimizer/optimizer_execution_frame.cc [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  // Create MLValues related maps
+  auto initialize_maps = [this, &initialized_tensor_set](const NodeArg* arg) -> Status {
+    int idx = mlvalue_name_idx_map_.Add(arg->Name());
+    mlvalue_idx_nodearg_map_[idx] = arg;
+
+    // Only create MLValue instances for initializers used by an array of nodes.
+    InitializedTensorSet::const_iterator it = initialized_tensor_set.find(arg->Name());
+    if (it != initialized_tensor_set.cend()) {
+      MLValue mlvalue;
+      utils::TensorProtoToMLValue(*(it->second), allocator_ptr_, nullptr, 0, mlvalue);
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/optimizer_execution_frame.cc]: >TensorProtoToMLValue [](start = 13, length = 20)

Not sure when the optimizer is running and what's available. Would the MLValue's for the initializers from the SessionState for the full Graph be available?

Comment (weixingzhang) [reply onnxruntime/core/optimizer/optimizer_execution_frame.cc]: It is running after the model is loaded and before SessionState is fully initialized. MLValues for the initializers from the SessionState are not available.

File: onnxruntime/core/optimizer/optimizer_execution_frame.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+#include "core/graph/graph.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+#include "core/framework/execution_frame.h"
+#include "core/framework/mlvalue_name_idx_map.h"
+#include "core/framework/ml_value.h"
+
+namespace onnxruntime {
+
+class OptimizerExecutionFrame final : public IExecutionFrame {
```
Comment (kkaranasos) [inline onnxruntime/core/optimizer/optimizer_execution_frame.h]: The PR looks good, thanks for the work @weixingzhang.
I have two high-level comments:
- Does the CPU execution provider have to be hardcoded? I was thinking that we could eventually use other providers in this execution frame.
- Would this streamlined version of the execution frame be useful outside the optimizer? If so, we could give it a more generic name than OptimizerExecutionFrame.

Comment (weixingzhang) [reply onnxruntime/core/optimizer/optimizer_execution_frame.h]: Thanks for the feedback.

The reason to use CPU execution provider is that only CPU execution provider supports all of operations with all possible inputs format. I was thinking to define Info ctor with an parameter of an instance of execution provider, that means the user of Info needs to create execution provider, which I think it is not necessary for now. Instead, it is better to centralizer the creation of execution provider.

Regarding to OptimizerExecutionFrame, I would like not to make it generic until it is really useful for other cases. Also, at that time, probably the code related to ExecutionFrame needs to be refactored again.

**Pull request merged**